pkg/cloud/services/compute/instance.go

```diff
@@ -490,16 +502,31 @@ func (s *Service) getBlockDevices(eventObject runtime.Object, instanceSpec *Inst
 	}
 
 	for _, blockDeviceSpec := range instanceSpec.AdditionalBlockDevices {
-		blockDevice, err := s.getOrCreateVolumeBuilder(eventObject, instanceSpec, blockDeviceSpec, "", fmt.Sprintf("Additional block device for %s", instanceSpec.Name))
-		if err != nil {
-			return []bootfromvolume.BlockDevice{}, err
+		var bdUUID string
+		var sourceType bootfromvolume.SourceType
+		var destinationType bootfromvolume.DestinationType
+		if blockDeviceSpec.Storage.Type == infrav1.VolumeBlockDevice {
+			blockDevice, err := s.getOrCreateVolumeBuilder(eventObject, instanceSpec, blockDeviceSpec, "", fmt.Sprintf("Additional block device for %s", instanceSpec.Name))
+			if err != nil {
+				return []bootfromvolume.BlockDevice{}, err
+			}
+			bdUUID = blockDevice.ID
+			sourceType = bootfromvolume.SourceVolume
+			destinationType = bootfromvolume.DestinationVolume
+		} else if blockDeviceSpec.Storage.Type == infrav1.LocalBlockDevice {
+			sourceType = bootfromvolume.SourceBlank
+			destinationType = bootfromvolume.DestinationLocal
+		} else {
+			return []bootfromvolume.BlockDevice{}, fmt.Errorf("invalid block device type %s", blockDeviceSpec.Storage.Type)
 		}
+
 		blockDevices = append(blockDevices, bootfromvolume.BlockDevice{
-			SourceType:          bootfromvolume.SourceVolume,
-			DestinationType:     bootfromvolume.DestinationVolume,
-			UUID:                blockDevice.ID,
+			SourceType:          sourceType,
+			DestinationType:     destinationType,
+			UUID:                bdUUID,
 			BootIndex:           -1,
 			DeleteOnTermination: true,
+			VolumeSize:          blockDeviceSpec.SizeGiB,
```

Comment:
We shouldn't set this unconditionally here because it doesn't apply to volumes. Hurray for unit tests!